### PR TITLE
[poc] add conditional resources to the static resource controller

### PR DESCRIPTION
**Pull request by deads2k**
@ravisantoshgudimetla had a PR very similar to this with some other stuff layered on top.
@ingvagabund had a PR that has a specific implementation of this for PDBs

I think this is a basis for PDB creation on some topologies and not others by using conditionals.

/assign @ingvagabund

### CODE REVIEW

File: pkg/operator/staticresourcecontroller/static_resource_controller.go [hunk 1]
```diff
@@ -67,6 +66,23 @@ type StaticResourceController struct {
 	categoryExpander restmapper.CategoryExpander
 }
 
+// ConditionalFunction provides needed dependency for a resource on another condition instead of blindly creating
+// a resource. This conditional function can also be used to delete the resource when not needed.
+type ConditionalFunction func() bool
```
Comment (ingvagabund) [inline pkg/operator/staticresourcecontroller/static_resource_controller.go:71]: Already there: https://github.com/openshift/library-go/blob/master/pkg/operator/resource/resourceapply/generic.go#L58-L60

File: pkg/operator/staticresourcecontroller/static_resource_controller.go [hunk 2]
```diff
@@ -213,14 +266,34 @@ func (c StaticResourceController) Sync(ctx context.Context, syncContext factory.
...
+		shouldCreate := conditionalManifest.shouldCreateFn()
+		shouldDelete := conditionalManifest.shouldDeleteFn()
+
+		var directResourceResults []resourceapply.ApplyResult
+
+		switch {
+		case !shouldCreate && !shouldDelete:
+			// no action required
+			continue
+		case shouldCreate && shouldDelete:
```
Comment (ingvagabund) [inline pkg/operator/staticresourcecontroller/static_resource_controller.go:279]: Worth writing a unit test at least for this case.

File: pkg/operator/staticresourcecontroller/static_resource_controller.go [hunk 3]
```diff
@@ -106,76 +121,114 @@ func (c *StaticResourceController) WithIgnoreNotFoundOnCreate() *StaticResourceC
...
 
+// WithConditionalResources adds a set of manifests to be created when the shouldCreateFnArg is true and should be
+// deleted when the shouldDeleteFnArg is true.
+// If shouldCreateFnArg is nil, then it is always create.
+// If shouldDeleteFnArg is nil, then it is !shouldCreateFnArg
+//  1. shouldCreateFnArg == true && shouldDeleteFnArg == true - produces an error
+//  2. shouldCreateFnArg == false && shouldDeleteFnArg == false - does nothing as expected
+//  3. shouldCreateFnArg == true applies the manifests
+//  4. shouldDeleteFnArg == true deletes the manifests
+func (c *StaticResourceController) WithConditionalResources(manifests resourceapply.AssetFunc, files []string, shouldCreateFnArg, shouldDeleteFnArg ConditionalFunction) *StaticResourceController {
```
Comment (ingvagabund) [inline pkg/operator/staticresourcecontroller/static_resource_controller.go:132]: Worth unit testing this for nil create/delete condition functions they get set to default always true/false.

Possibly reuse https://github.com/openshift/library-go/pull/1110/files#diff-2ba2b98cfb79db1510fd030672e638dac32c02bce52f8238b5c10651fb7ded8d ?

**Review by ingvagabund — CHANGES_REQUESTED** [review]
Just few nits

**Comment by deads2k** [comment]
```
[deads@fedora library-go]$ git push -f
Enumerating objects: 797, done.
Counting objects: 100% (495/495), done.
Delta compression using up to 12 threads
Compressing objects: 100% (132/132), done.
Writing objects: 100% (337/337), 136.92 KiB | 34.23 MiB/s, done.
Total 337 (delta 212), reused 311 (delta 194), pack-reused 0
remote: Resolving deltas: 100% (212/212), completed with 101 local objects.
To github.com:/deads2k/library-go
 ! [remote rejected]     conditional -> conditional (cannot lock ref 'refs/heads/conditional': reference already exists)
error: failed to push some refs to 'github.com:/deads2k/library-go'

```

hmmmm.... problem for tomorrow I guess.

**Pull request merged**